lib/resizer.js

```diff
@@ -40,85 +43,118 @@ exports.work = (imgEvent) => {
-  }
+exports.upload = (tmpFile, destPath, contentType) => {
+  let fileName = path.basename(tmpFile);
   return Q.ninvoke(s3, 'upload', {
-    Body: fileStream,
+    Body: fs.createReadStream(tmpFile),
+    Bucket: process.env.DESTINATION_BUCKET,
+    Key: `${destPath}/${fileName}`,
+    ContentType: contentType
+  });
+}
+
+// copy between s3 locations
+exports.copy = (tmpBucket, tmpKey, destPath, contentType) => {
+  let fileName = path.basename(tmpKey);
+  return Q.ninvoke(s3, 'copyObject', {
+    CopySource: `${tmpBucket}/${tmpKey}`,
     Bucket: process.env.DESTINATION_BUCKET,
-    Key: `${filePath}/${fileName}`
+    Key: `${destPath}/${fileName}`,
+    MetadataDirective: 'REPLACE',
+    ContentType: contentType
   });
 }
 
 // resize thumbs, then move to the destination
-exports.resizeAll = (fileBuffer, filePath, fileName) => {
+exports.resizeAll = (tmpFile, tmpS3Bucket, tmpS3Key, destPath, contentType) => {
   return Q.all(RESIZES.map(size => {
-    if (size.key === 'original') {
-      return exports.upload(fileBuffer, filePath, fileName);
+    if (size.key === 'original' && tmpS3Bucket && tmpS3Key) {
+      return exports.copy(tmpS3Bucket, tmpS3Key, destPath, contentType);
+    } else if (size.key === 'original') {
+      return exports.upload(tmpFile, destPath, contentType);
     } else {
-      return exports.resize(fileBuffer, size.args).then(resizedStream => {
-        return exports.upload(resizedStream, filePath, fileName, size.key);
+      return exports.resize(tmpFile, size).then(data => {
+        return exports.upload(data.path, destPath, contentType);
       });
     }
   }));
 }
 
-// resize a file, returning a stream
-exports.resize = (fileBuffer, resizeArgs) => {
-  let magicker = gm(fileBuffer);
-  let sizer = magicker.resize.apply(magicker, resizeArgs);
-  return Q.ninvoke(sizer, 'stream').then(results => results[0]);
+// resize a file, returning metadata about the file
+exports.resize = (tmpFile, size) => {
+  let resize = sharp(tmpFile).resize(size.width, size.height);
+  if (size.exact) {
+    resize = resize.crop(sharp.strategy.entropy);
+  } else {
+    resize = resize.max();
+  }
+
+  let parts = path.parse(tmpFile);
+  let dest = os.tmpdir() + `/${parts.name}_${size.key}${parts.ext}`;
+  return resize.toFile(dest).then(data => {
+    data.path = dest;
+    return data;
+  });
 }
 
 // identify and validate a file
-exports.validate = (fileBuffer) => {
-  return Q.ninvoke(gm(fileBuffer), 'identify').then(
-    info => {
-      let meta = {
-        width: info.size.width,
-        height: info.size.height,
-        filesize: fileBuffer.length,
-        format: info['format'].toLowerCase()
+exports.validate = (tmpFile) => {
+  let stat = Q.ninvoke(fs, 'stat', tmpFile);
+  let info = sharp(tmpFile).metadata();
+  return Q.all([stat, info]).then(
+    res => {
+      return {
+        width: res[1].width,
+        height: res[1].height,
+        filesize: res[0].size,
```

Comment:
Oddly, sharp's output included a file size.  But it was always wrong.  Maybe because it's stripping exif data, even on a `metadata` call?